src/EditorFeatures/CSharpTest/ForeachToFor/ForeachToForTests.cs

```diff
@@ -0,0 +1,1177 @@
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (/* test */ var a in array) ;
+    }
+}
+";
+            await TestMissingInRegularAndScriptAsync(text);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task CommentNotSupported3()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (var a in array /* test */) ;
+    }
+}
+";
+            await TestMissingInRegularAndScriptAsync(text);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task CollectionStatement()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        foreach[||](var a in new int[] { 1, 3, 4 })
+        {
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var {|Rename:list|} = new int[] { 1, 3, 4 };
+        for (var {|Rename:i|} = 0; i < list.Length; i++)
+        {
+            var a = list[i];
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task CollectionConflict()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var list = 1;
+
+        foreach[||](var a in new int[] { 1, 3, 4 })
+        {
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var list = 1;
+
+        var {|Rename:list1|} = new int[] { 1, 3, 4 };
+        for (var {|Rename:i|} = 0; i < list1.Length; i++)
+        {
+            var a = list1[i];
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task IndexConflict()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (var a in array)
+        {
+            int i = 0;
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        for (var {|Rename:i1|} = 0; i1 < array.Length; i1++)
+        {
+            var a = array[i1];
+            int i = 0;
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task VariableWritten()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (var a in array)
+        {
+            a = 1;
```

Comment:
as far as I know, this code doesn't actually compile (foreach variable can't be written to). what's the reason for this?

Reply:
we offer refactoring for code that doesn't compile as well. I wrote the reason above.

Reply:
> we offer refactoring for code that doesn't compile as well

yes, that's what I would expect, but this test actually prevents that

Reply:
offering refactoring on broken code and refactoring changing meaning of code is different thing. refactoring is not supposed to change meaning of code.

Reply:
> refactoring is not supposed to change meaning of code

actually the vast majority of refactorings do this

Reply:
In general, the only rule we really have here is that a 'large scale refactoring' really should preserve semantics.  "large scale" means "large enough that's its unreasonable for a user to have to examine the code to validate that the refactoring didn't subtly change anything.  Basically, the only examples of this are "rename" and, to a lesser extent, "replace prop with methods (and vice versa).  

Almost everything else we have is not really 'large scale', and is much more of a "locally targeted code transformation".  In that regard we have a *lot* more leeway with what we offer.  And we can definitely still be offered even in broken code scenarios.

Reply:
sure. everyone has their reasoning. I opened an issue tracking all design questions or handling additional cases.

https://github.com/dotnet/roslyn/issues/25667

Reply:
> should refactoring be offered even if refactoring can change semantic

I don' tthink this should even be shown as a warning, I don't really see this as "changing the semantics". Yes, if you're trying to modify the variable, the code is invalid, But the resulting code would have the exact same semantics as reassigning the foreach variable if it was allowed. The fact that you can't modify the foreach variable is just an arbitrary limitation. There's no semantics to be concerned with.

Reply:
I am fine with going permissive way. 

if we decide refactoring is more about reducing work for code transformation, then I would go even more than this actually. like there is no reason why I dont offer refactoring because type or variable, or collection is missing. or because next statement's control variable is different than foreach variable and etc.

I dont even need to check whether length or indexer is there, I can offer the refactoring and do code transformation. 

still less work for users to fix it up later.